tls/s2n_connection.c

```diff
@@ -138,6 +144,49 @@ static int s2n_connection_wipe_keys(struct s2n_connection *conn)
     return 0;
 }
 
+static int s2n_connection_free_io_contexts(struct s2n_connection *conn)
+{
+    /* Free the I/O context if it was allocated by s2n. Don't touch user-controlled contexts. */
+    if (!conn->managed_io) {
+        return 0;
+    }
+    
+    struct s2n_blob send_io_blob;
+    struct s2n_blob recv_io_blob;
+
+    if (conn->send_io_context) {
+        send_io_blob.data = (uint8_t *)conn->send_io_context;
+        send_io_blob.size = sizeof(struct s2n_socket_write_io_context);
+        GUARD(s2n_free(&send_io_blob));
+    }
+
+    if (conn->recv_io_context) {
+        recv_io_blob.data = (uint8_t *)conn->recv_io_context;
+        recv_io_blob.size = sizeof(struct s2n_socket_read_io_context);
+        GUARD(s2n_free(&recv_io_blob));
+    }
+
+    return 0;
+}
+
+static int s2n_connection_wipe_io(struct s2n_connection *conn)
+{
+    if (conn->managed_io && conn->recv){
```

Comment:
Good call out. I don't foresee a situation where a user would want managed read but non-managed write (or vice-versa). Worth documenting. We could also update the API to make this explicit. Something like `s2n_connection_set_io_cb(s2n_connection *conn, s2n_recv_fn recv, s2n_send_fn send)` ?